server/auth/auth.go

```diff
@@ -0,0 +1,740 @@
+// Copyright 2020 TiKV Project Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissionKeys and
+// limitations under the License.
+
+package auth
+
+import (
+	"encoding/json"
+	"github.com/pingcap/log"
+	"github.com/tikv/pd/pkg/errs"
+	"go.uber.org/zap"
+	"path"
+	"strings"
+	"sync"
+
+	"github.com/pingcap/errors"
+	"github.com/tikv/pd/server/kv"
+	"go.etcd.io/etcd/clientv3"
+)
+
+var (
+	// ErrUserNotFound is error info for user not found.
+	ErrUserNotFound = func(name string) error { return errors.Errorf("user not found: %s", name) }
+	// ErrUserExists is error info for user exists.
+	ErrUserExists = func(name string) error { return errors.Errorf("user already exists: %s", name) }
+	// ErrRoleNotFound is error info for role not found.
+	ErrRoleNotFound = func(name string) error { return errors.Errorf("role not found: %s", name) }
+	// ErrRoleExists is error info for role exists.
+	ErrRoleExists = func(name string) error { return errors.Errorf("role already exists: %s", name) }
+	// ErrUserHasRole is error info for user has role.
+	ErrUserHasRole = func(user string, role string) error { return errors.Errorf("user %s already has role: %s", user, role) }
+	// ErrUserMissingRole is error info for user lacks a required role.
+	ErrUserMissingRole = func(user string, role string) error {
+		return errors.Errorf("user %s doesn't have role: %s", user, role)
+	}
+	// ErrRoleHasPermission is error info for role has permission.
+	ErrRoleHasPermission = func(role string, permission Permission) error {
+		return errors.Errorf("role %s already has permission: %s", role, permission.String())
+	}
+	// ErrRoleMissingPermission is error info for role lacks a required role.
+	ErrRoleMissingPermission = func(role string, permission Permission) error {
+		return errors.Errorf("role %s doesn't have permission: %s", role, permission.String())
+	}
+)
+
+var (
+	userPrefix = "users"
+	rolePrefix = "roles"
+)
+
+// Manager is used for the rbac storage, cache, management and enforcing logic.
+type Manager struct {
+	kv    kv.TxnBase
+	users sync.Map // map[string]*User
+	roles sync.Map // map[string]*Role
+}
```

Comment:
I found it all the methods related to `rbac control` is managed by this Manager which makes it so large. As I recommended [here](https://github.com/tikv/pd/pull/3224/files#r533090691). We could name it as `rbacManager` use which contains userManager and roleManager.

The `userManger` manages the methods for the `users` including `querying from the memory` / `loading from the etcd` / `persisting into etcd` / `comparing the password` etc...

The `roleManager ` manages the methods for the `roles`  including `querying from the memory` / `loading from the etcd` / `persisting into etcd` etc...